File: 2312-selling-pieces-of-wood/2312-selling-pieces-of-wood.cpp
class Solution {
public:
    long long sellingWood(int m, int n, vector<vector<int>>& prices) {
        long long dp[201][201] = {};
        for(auto price:prices){
            dp[price[0]][price[1]] = price[2];
        }
        for(int row = 1; row <= m ; row++){
            for(int col = 1; col <= n ; col++){
                for(int cut_at = 1; cut_at<=(col/2); cut_at++){
                    dp[row][col] = max(dp[row][col],dp[row][cut_at]+dp[row][col-cut_at]);
                }
                for(int cut_at = 1; cut_at<=(row/2); cut_at++){
                    dp[row][col] = max(dp[row][col],dp[cut_at][col]+dp[row-cut_at][col]);
                }
            }
        }
        return dp[m][n];
    }
};

File: 2312-selling-pieces-of-wood/2312-selling-pieces-of-wood.py
class Solution:
    def sellingWood(self, m: int, n: int, prices: List[List[int]]) -> int:
        dp = [[0 for _ in range(n+1)] for _ in range(m+1)]
        for w, h, p in prices:
            dp[w][h] = p
        for w in range(1, m + 1):
            for h in range(1, n + 1):
                for a in range(1, w // 2 + 1):
                    dp[w][h] = max(dp[w][h], dp[a][h] + dp[w - a][h])
                for a in range(1, h // 2 + 1):
                    dp[w][h] = max(dp[w][h], dp[w][a] + dp[w][h - a])
        return dp[m][n]
        